File: BOJ/java/10974.java
/*
 * 모든 순열
 * https://www.acmicpc.net/problem/10974
 */

import java.io.IOException;
import java.util.Scanner;

class Main {
    static int n;
    static int[] a;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        a = new int[n];
        sc.close();
        for (int i = 0; i < n; i++) {
            a[i] = i + 1;
        }

        do {
            for (int i : a) sb.append(i).append(" ");
            sb.append("\n");
        } while (nextPermutation());
        System.out.println(sb.toString());
    }

    static boolean nextPermutation() {
        int lastIndex = a.length - 1;
        int i = lastIndex;
        while (i > 0 && a[i] < a[i - 1]) {
            i--;
        }
        if (i == 0) return false;

        int j = lastIndex;
        while (j >= i && a[j] < a[i - 1]) {
            j--;
        }
        swap(i - 1, j);

        j = lastIndex;
        while (i < j) swap(i++, j--);
        return true;
    }

    static void swap(int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
}

File: BOJ/kotlin/10974.kt
import java.lang.StringBuilder

fun main() {
    val n = readLine()!!.toInt()
    val a = (1..n).toMutableList()
    val sb = StringBuilder()

    do {
        sb.append(a.joinToString(" ")).append("\n")
    } while (nextPermutation(a))
    println(sb)
}

fun nextPermutation(a: MutableList<Int>): Boolean {
    val lastIndex = a.size - 1
    var i = (lastIndex downTo 1).find { a[it] > a[it - 1] } ?: return false
    var j = (lastIndex downTo i).find { a[it] > a[i - 1] } ?: return false
    a.swap(i - 1, j)
    j = lastIndex
    while (i < j) a.swap(i++, j--)
    return true
}

fun <T> MutableList<T>.swap(i: Int, j: Int) {
    val temp = this[i]
    this[i] = this[j]
    this[j] = temp
}